### PR TITLE
Add a SafeHttpClient method

### PR DESCRIPTION
**Pull request by biilmann**
This adds a tracing method to a client that will make sure
any request to a reserved IP is canceled at DNSDone time

### CODE REVIEW

File: http/http_test.go [hunk 1]
```diff
@@ -14,3 +17,40 @@ func TestIsLocalAddress(t *testing.T) {
...
+
+func (c *countServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	c.count++
+	w.Write([]byte("Done"))
+}
+
+func TestSafeHTTPClient(t *testing.T) {
+	client := &http.Client{}
+
+	counter := &countServer{}
```
Comment (rybit) [inline http/http_test.go]: I'd use the `httptest` lib instead. https://golang.org/pkg/net/http/httptest/#Server

File: http/http.go [hunk 2]
```diff
@@ -49,4 +52,48 @@ func SafeDialContext(dialContext func(ctx context.Context, network, addr string)
...
+	for _, addr := range info.Addrs {
+		fmt.Printf("Checking addr: %v\n", addr)
+		if isPrivateIP(addr.IP) {
+			fmt.Printf("Got private IP %v\n", addr.IP)
+			t.cancel()
+		}
+	}
+}
+
+func SafeHttpClient(client *http.Client) *http.Client {
```
Comment (rybit) [inline http/http.go]: s/Http/HTTP

**Comment by biilmann** [comment]
I like that suggestion - plopped that implementation in there

File: http/http.go [hunk 3]
```diff
@@ -35,18 +39,58 @@ func isPrivateIP(ip net.IP) bool {
...
-	return func(ctx context.Context, network, addr string) (net.Conn, error) {
-		if isLocalAddress(addr) {
-			return nil, errors.New("Connection to local network address denied")
-		}
+func (no noLocalTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	ctx, cancel := context.WithCancel(req.Context())
+
+	ctx = httptrace.WithClientTrace(ctx, &httptrace.ClientTrace{
+		ConnectStart: func(network, addr string) {
+			fmt.Printf("Checking network %v\n", addr)
```
Comment (mraerino) [inline http/http.go]: This should be removed because it pollutes logs

File: http/http.go [hunk 4]
```diff
@@ -35,18 +39,58 @@ func isPrivateIP(ip net.IP) bool {
...
+			ip := net.ParseIP(host)
+			if ip == nil {
+				cancel()
+				fmt.Printf("Canceleing dur to error in ip parsing %v", host)
+				return
+			}
+
+			if isPrivateIP(ip) {
+				cancel()
+				fmt.Println("Canceleing dur to private ip range")
```
Comment (mraerino) [inline http/http.go]: would be great to log these and previous errors on the `errlog` field of the noLocalTransport

**Comment by vbrown608** [comment]
I'm going to update the tests a bit

**Comment by mraerino** [comment]
There also is a typo in `SafeRountripper` you know? 😛

**Comment by vbrown608** [comment]
Ah, I thought you intended for that to be a follow-up (along with un-inlining this functionality in gotrue). I can put it in here.

**Pull request merged**